### PR TITLE
Config manager

### PR DESCRIPTION
**Pull request by neeleshdodda44**
Sorry about the delay. Should be a straight forward pull request. Couple things (quoting the Trello card):

> These "question entries" can be stored as a JSON in `core/configuration/questions.json`

Honestly I thought this was a textbook description of a CSV since every "row" has a question, , config section, key and default answer. So I just made those the columns and created a CSV. I still think it's easy to update, since adding to the question bank is simply adding a row. 

> Check setup mode is entered when the configuration file doesn't exist

Yes, I know `configuration.ini` exists right now. But the tests **remove** the file (using `reset`) at the beginning, so a new config is actually created for each test. Why don't I just use a new directory for testing? I could mock out `ConfigurationManager` in the components that depend on it, but I ultimately decided that it was important to test that each file can access the config in their respective ways. Because each test removes `configuration.ini` at the start and leave a copy of the default `configuration.ini` at the end, the functionality is still proven to be correct.

### CODE REVIEW

File: tests/test_configurationmanager.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+import tests.context
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply tests/test_configurationmanager.py]: 1. I disagree. How do we test `run_setup_mode` without creating a new config file?

2. Actually (might just be a pytest thing) creating a new instance for each test triggers the singleton exception. But yes, I can just use fixtures.

3. Agreed. Would probably accomplish this with fixtures, if I understand them correctly.

[the remark this answers is not in the document]
Comment (georgymh) [reply tests/test_configurationmanager.py]: 1. There is no need to delete the config file outside of testing. To test `run_setup_mode` you could set `config_filepath` to another path inside of the artifacts folder and clean up right after the test. Since this is only relevant to testing, this clean up should be part of the tests (not the class).

2. Does this happen even if you call `getInstance()` in the fixture?

[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply tests/test_configurationmanager.py]: 1. I think I see what you're saying now. So here's what I'm thinking:
* Delete `core/configuration.ini` (since it's only needed at the start for testing).
* Create `tests/artifacts/configuration.ini` so that every other test can use that for testing.
* Create `tests/artifacts/config_manager` so that the configuration manager can create config files in this folder for testing (and the folders will be removed as a part of testing, not a method that's a part of `ConfigurationManager`.

2. No, since a new instance would not be created each test. So yes, we can use `getInstance()` in the fixture.

I'll update this thread when I make the necessary changes.

[the remark this answers is not in the document]
Comment (georgymh) [reply tests/test_configurationmanager.py]: Sounds good.

File: core/configuration.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+			self.config = None
+
+	def bootstrap(self, test_input=None):
+		""" Do nothing if config instance already exists. Otherwise, check if core/configuration.ini exists. If not, call 
+			run_setup_mode. Then load the config instance from configuration.ini
+
+			'test_input' is just an easy way to mock out user input for testing """
+		if not self.config:
+			user_input = test_input if test_input else input
+			print(os.listdir('core'))
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply core/configuration.py]: Done.

File: core/configuration.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+			if not os.path.isfile(self.config_filepath):
+				print("found")
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply core/configuration.py]: Done.

File: core/configuration.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+			question, section, key, default = row['question'], row['section'], row['key'], row['default']
+			answer = user_input(self.question_format.format(question=question, default=default))
+			answer = answer if answer else default
+			if not section in config:
+				config[section] = {}
+			config[section][key] = answer
+		with open(self.config_filepath, 'w') as configfile:
+			config.write(configfile)
+
+	def reset(self):
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply core/configuration.py]: Removed.

File: core/dataset_manager.py [hunk 5]
```diff
@@ -80,11 +81,16 @@ class DatasetManager():
             06/10/16mlf.csv
 
     """
-    def __init__(self, raw_filepath):
+    def __init__(self, raw_filepath = None):
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply core/dataset_manager.py]: Done. Was trying not to rewrite tests, but then I realized it was a quick change so I went ahead and did it.

File: core/configuration.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+		""" Virtually private constructor. """
+		if ConfigurationManager.__instance != None:
+			raise Exception("This class is a singleton!")
+		else:
+			ConfigurationManager.__instance = self
+			self.config_filepath = config_filepath
+			self.question_format = "{question} [default = {default}] "
+			self.config = None
+
+	def bootstrap(self, test_input=None):
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply core/configuration.py]: Done.

File: tests/test_configurationmanager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply tests/test_configurationmanager.py]: Made these changes. Decided that `getInstance` (now `get_instance to keep everything camel case) should take parameters updating the current config file location and input function (which differ among tests).

File: core/configuration.py [hunk 7]
```diff
@@ -12,13 +12,19 @@ class ConfigurationManager(object):
 	__instance = None
 
 	@staticmethod
-	def getInstance():
-		""" Static access method. """
-		if ConfigurationManager.__instance == None:
-			ConfigurationManager()
+	def get_instance(config_filepath="core/configuration.ini", input_function=None):
```
Comment (georgymh) [inline core/configuration.py]: So the design change of having these two parameters on the `get_instance()` function introduces a couple problems:

1. Whenever we use a `config_filepath` other than the default, each consumer of the Configuration Manager needs to know, store, and pass around this variable (which makes it cumbersome and exposes test-related variables and logic across the entire codebase).

2. Bootstrap is called **not only** once during initialization, but every time during the retrieval of the singleton (which is inefficient).

To solve these, I propose we make the `bootstrap()` function take in the `config_filepath` and `input_function`, the `get_instance()` function take no parameters, and that we make a getter for the `config` property that raises an exception if the class hasn't been bootstrapped yet.

At the end of the day we just need to bootstrap the Configuration Manager only once during the very beginning of the service, so this works in accordance to that. For tests, just bootstrap right after you call `get_instance()` for the first time (or every time you call the instance getter if you're creating new CM's for every single test).

File: tests/test_configurationmanager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (georgymh) [reply tests/test_configurationmanager.py]: It seems we were working in parallel, but I made some comments in regards to the new get_instance. I think you need to go to "file view" to see them.

File: core/configuration.py (hunk 7 above)
Comment (neeleshdodda44) [reply core/configuration.py]: _Summarizing what I said in meeting for clarity_

I revamped the design of the class. `__init__` takes optional parameters for `config_filepath` and `input_function`, which are attributes of the instance. `bootstrap` and `get_instance` take no parameters. So from a consumer's point of view (as long as the file path does not change), all that needs to be done is initialize the class and call `get_instance` to retrieve the instance.

The tests are a bit more complex. So in order to uphold all the design suggestions you made (singleton, keep `bootstrap` in `__init__`, don't use test variables, etc.) I decided to use fixtures. Each test file creates/alters a `ConfigurationManager` instance so that it knows where to pull/create the config file. The only exception is `DMLScheduler`, where all of that happens outside of the test methods and fixtures because the runners never finish running otherwise. Not sure if this worth worrying about for now, but I can create a TODO card to try to debug why fixtures don't work for the scheduler at a later point. 

Tl;dr tests pass and from a consumer standpoint, I believe I've addressed all your concerns.

**Comment by neeleshdodda44** [comment]
_Summarizing what I said in meeting for clarity_

I revamped the design of the class. `__init__` takes optional parameters for `config_filepath` and `input_function`, which are attributes of the instance. `bootstrap` and `get_instance` take no parameters. So from a consumer's point of view (as long as the file path does not change), all that needs to be done is initialize the class and call get_instance to retrieve the instance.

The tests are a bit more complex. So in order to uphold all the design suggestions you made (singleton, keep `bootstrap` in `__init__`, don't use test variables, etc.) I decided to use fixtures. Each test file creates/alters a `ConfigurationManager` instance so that it knows where to pull/create the config file. The only exception is `DMLSchedule`r, where all of that happens outside of the test methods and fixtures because the runners never finish running otherwise. Not sure if this worth worrying about for now, but I can create a TODO card to try to debug why fixtures don't work for the scheduler at a later point.

Tl;dr tests pass and from a consumer standpoint, I believe I've addressed all your concerns.

File: core/runner.py [hunk 8]
```diff
@@ -29,13 +29,11 @@ class DMLRunner(object):
 
     """
 
-    def __init__(self):
+    def __init__(self, config_manager):
```
Comment (neeleshdodda44) [inline core/runner.py]: So I see that now we're passing the config manager instance and calling `reset`. I like the change, since it makes the tests a lot cleaner at the expense of little for the actual classes. 

But I wanted to ask (nitpick): if we're going to go as far as to pass the instance around, can we just make the config manager not a singleton? It feels like we're passing around a config file where each class uses a different part of the config. Why don't we just have separate configs for separate classes? For example, `DMLRunner` could have `configuration_runner.ini`, and so forth. Then each class creates its own instance of the config manager with respect to its own config file. I feel like testing in the future would significantly easier, depending on how many more components of the Unix Service require their own set of configurations.

Comment (georgymh) [reply core/runner.py]: I agree with you. If we're doing Dependency Injection we should probably not do Singleton too, as it opens up 2 ways of accessing the class: by passing it around and by calling the get instance static method.

The benefit of DI over Singleton is that testing is a lot easier and probably readability too, so I say let's stick to DI, make the class a non-Singleton, and make the fact that there is only 1 Configuration Manager an implementation detail.

In regards to having separate configurations, I don't really see a benefit there. The idea of having only 1 configuration with multiple sections adds more ease of use and is less complex in my opinion.

Now, this makes me reconsider the design choice of making Dataset Manager a Singleton. Maybe we should change it and make the (future) Listener inject the DM...

Comment (neeleshdodda44) [reply core/runner.py]: Okay, the class is not a singleton anymore. Since there's still only 1 Configuration Manager, I still think it makes sense to use `get_instance`. So tl;dr just removed the enforcement of only one instance.

Totally agree that DI is easier. Especially this case, since we're using the Configuration Manager in basically three different ways. 

Inject the DM in the Scheduler -> Runners, yes? That makes sense, but we can make an official decision once we work on the Listener.

If everything looks good, I can go ahead and merge?

Comment (georgymh) [reply core/runner.py]: Let's go ahead and remove the `get_instance()` and `_instance` property so we don't get confused thinking the class is a Singleton in the future. This also comes with changing the tests to initialize the class normally. After that we're ready to merge.

Comment (neeleshdodda44) [reply core/runner.py]: Sorry about the delay. Made the changes you asked for and I will be merging now.

**Comment by neeleshdodda44** [comment]
Looks like we've come into agreement with the final design of ConfigurationManager. Merging now.
